### ISSUE TITLE
Recommend/encourage developers to build for Linux 64-bit

### ISSUE THREAD

**Issue by wrsg** [itchio/itch.io#543]
This affects the Unity games especially, many only package 32-bit builds for some reason. 
I don't know the current metrics but I don't think many people who want to play games on Linux are running on pre-64-bit hardware.

**Comment by felixplesoianu** [comment]
I am. And generally speaking, people build for the systems they have. Not everyone can afford hardware from this decade, let alone this year.

**Comment by leafo** [comment]
@felixplesoianu many 64 bit users do not have the required multiarch packages installed to run them. It's more likely they'll give up installing the game then go and install the required decencies. I think it's important to educate uploaders on choosing how to build their unity games.

Does anyone have the unity editor installed on windows, can you provide images on how you can configure the build that I can use in a guide on the itch.io website?

**Comment by fasterthanlime** [comment]
> I am. And generally speaking, people build for the systems they have. Not everyone can afford hardware from this decade, let alone this year. And 64-bit Linux often still has library issues, so people may well have reasons to stick with 32-bit even if they have the hardware. And these are just two reasons off the top of my head.

I don't think the argument is to drop 32-bit binaries (some folks are arguing for that, but not me (and I don't believe @wrsg is either)).

There's a small but vocal group that straight up doesn't bother with 64-bit releases (hi @Cheeseness o/), I'm of the opinion that developers should release both so that as many people as possible can play their games.

**Comment by Cheeseness** [comment]
> There's a small but vocal group that straight up doesn't bother with 64-bit releases (hi @Cheeseness o/), I'm of the opinion that developers should release both so that as many people as possible can play their games.

Sorry, what? I have always maintained that supporting both being the ideal for me. The majority of the games I have available on itch.io have 64 bit builds available (some don't even have 32 bit builds right now).

To clarify, I believe that for my own projects, if I don't have the opportunity to ship both 32 and 64 bit builds, then the best fallback that gives the biggest opportunity to support both architectures is to ship 32 bit builds. This belief will likely change if/when distros drop 32 bit hardware support, but 32 bit software support will be around for decades beyond that to allow my older titles to be playable for a reasonable lifespan.

> Does anyone have the unity editor installed on windows, can you provide images on how you can configure the build that I can use in a guide on the itch.io website?

I could probably grab some screenshots tomorrow of the Windows version of the Unity editor in Wine (I think the export dialogue is the same on the Linux version though?) if you can't find anybody in the meantime.

**Comment by fasterthanlime** [comment]
> Sorry, what? I have always maintained that supporting both is the ideal for me. The majority of the games I have available on itch.io have 64 bit builds available (some don't even have 32 bit builds right now).

Super happy to be wrong in that occurence! I might be thinking of someone else :)

**Comment by fasterthanlime** [comment]
> I could probably grab some screenshots tomorrow of the Windows version of the Unity editor in Wine (I think the export dialogue is the same on the Linux version though?) if you can't find anybody in the meantime.

Actually, we already have a guide that has a screenshot of that :) https://itch.io/docs/itch/integrating/platforms/linux.html

**Comment by felixplesoianu** [comment]
> I think it's important to educate uploaders on choosing how to build their unity games.

I see the core of my argument flew over everyone's head.

I don't *have* a 64-bit machine. I *can't* build for an architecture I don't have. It's not a "choice". And anecdotally, I know a lot of people who have machines just as old as mine, if not older. You can't "educate" them into being able to afford a new computer. Might as well insist that every game published an itch.io also comes with a Mac build, even though most people don't have a Mac, or access to a Mac, or the knowledge to build for Macs, or the time to learn.

Please keep itch.io friendly for hobbyists, students, and the less fortunate, too. Don't make us feel unwelcome all of a sudden.

**Comment by Cheeseness** [comment]
> Super happy to be wrong in that occurence! I might be thinking of someone else :) (Was mentioning you explicitly to give you a chance to state your current viewpoint, so it worked out great!)

Ah, it happens ^_^


> I don't have a 64-bit machine.

When I first saw this issue, I was tempted to say that so far as education goes, there's probably more to be gained in terms of positive user experience by putting effort into lobbying distro maintainers to make multiarch support simple/easy and/or pushing back against some of the not-quite-so-reality-based anti-32 bit narratives that are floating around.

I've heard a few people yammering like enabling multiarch involves having their legs broken, and from my own experiences, I have to see that as more a result of assumption/misinformation than any distros being *that* bad. When I shipped 32 bit builds of Day of the Tentacle, I had certain people (not pointing any fingers!) linking me to a mailing list post suggesting that 32 bit hardware support should be dropped in the future (which last time I looked didn't have any favourable responses) as though it were confirmation that 32 bit software support was going away. There's definitely room for more clarity about what the actual "problems" are, I think.


> I can't build for an architecture I don't have.

Certain tools (like Unity or Godot) use precompiled binary templates. I haven't checked, but I'm of the impression that with both of those, the resource bundles are arch agnostic, so a 32 bit based developer could potentially ship 64 bit builds in some situations - that might not be helpful for you specifically @felixplesoianu, but if the general topic here is helping people understand their options so they can make informed choices that suit their priorities, then that's probably a good piece of information to confirm.

**Comment by felixplesoianu** [comment]
> I've heard a few people yammering like enabling multiarch involves having their legs broken, and from my own experiences, I have to see that as more a result of assumption/misinformation than any distros being that bad.

I can't speak from the perspective of a Linux distro developer, but as a *user* I definitely had problems with some distributions -- and I mean modern, friendly distributions -- failing to run 32-bit software properly in the 64-bit edition (important software that's not easily built from source, and not available for just any architecture). Until that's resolved, you'll forgive me if simple caution pushed me to use a 32-bit OS even on a 64-bit machine.

Which, again, I don't have, and you can't "educate" me into being able to afford one. *Please* drop that angle because it's positively insulting.

> Certain tools (like Unity or Godot) use precompiled binary templates.

So does Ren'Py. Both the SDK and the builds it generates are universal 32/64-bit on Linux. Other engines I use... not so much. And there are more important considerations in choosing a game engine, especially for a hobbyist developer with limited resources.

**Comment by Cheeseness** [comment]
> Which, again, I don't have, and you can't "educate" me into being able to afford one. Please drop that angle because it's positively insulting.

Apologies if you felt my statements were suggesting that you needed educating or that you should be moving up to a 64 bit system. That was not my intention. I've not interacted with you before today and have no reason to believe that you're among those who I see as exaggerating the difficulties of running multiarch.

> And there are more important considerations in choosing a game engine, especially for a hobbyist developer with limited resources.

Most definitely, but in the context of this discussion rather than any individual developer's priorities, this information is important.

**Comment by quyse** [comment]
Unity's metrics on Linux usage: http://hwstats.unity3d.com/pc/os-linux.html
Steam's stats are much less informative, but may have some use too: http://store.steampowered.com/hwsurvey/?platform=linux
According to Unity's stats, 97% of Linux users are using 64-bit OS.

Now I don't think itch.io must encourage developers to build **only** for 64-bit Linux, because obviously those 3% users will not be able to run 64 bit build, and surely it's good to not discriminate them (after all Linux community is itself still a minority in today's gaming world). However it's definitely worth to show that if you ship 32-bit build only, 97% of users may have problems with it.

**Comment by fasterthanlime** [comment]
> I don't have a 64-bit machine. I can't build for an architecture I don't have

Yes you can :) It's possible to cross-compile for a 64-bit machine from a 32-bit machine, just like it's possible to cross-compile for Windows from Linux (via mingw-w64 and others), or even for Mac from Linux (although Apple seems to make that harder with every iteration of macOS) — **but**, it's not a priority for you, and that's okay! No one advocated "banning games that only have 32-bit builds".

I think the original spirit of the issue was this: there's game developers that aren't so much in love with the technical side of it — to them, game engines, platforms, etc. are just a means to an end: they want to create the game they have in mind and unleash it before an audience.

Those developers usually make pragmatic decisions: some of them run Windows because "it's what most people run and it came with my computer!", some of them pick Unity because "it seems real popular and there's tons of people I can ask in case it goes south!".

And then, eventually in their itch.io comments, someone brings up a Linux build — so they open the Unity editor and look at the options, and, knowing the Windows world, they do a 32-bit build, thinking both 64-bit and 32-bit Linux users will be able to run it (which is true on Windows, but not so much on Linux). Then they upload that and call it a day (and the [compatibility watchlist](https://github.com/itchio/itch-compatibility-watchlist) grows longer..)

What (I think) OP meant is that at some point in that process, there has to be some sort of paragraph somewhere on itch.io that they encounter and that tells them "listen, a majority of Linux users are going to have trouble running 32-bit only builds, it'd be cool if you could do 64-too!" and then have options "Upload 64-bit build", "Maybe later", "Never bother me about this again" — all valid choices.

Of course, the other big issue with getting your game running on most Linux boxes out there is shipping the proper libs (which, again, [we have a guide for in the itch docs](https://itch.io/docs/itch/integrating/platforms/linux.html)) — but it can be a lot of work. Usually a lot more work than for Windows, about the same amount of work as on macOS, except on macOS it's universally accepted that shipping the libs you use in your app bundle is the way to go, whereas in the Linux scene there's folks strongly against that.

**TL;DR** Linux is (still, in 2016) hard to ship for (as in it takes time and knowledge — but not special hardware). No solution is going to make everyone happy — but itch.io can and should help developers follow a "known-good path" that at least allows most Linux users to run their games (even if some of them will complain that it's not their favourite compromise).

**Comment by wrsg** [comment]
@fasterthanlime Just wondering if there is any progress on this becoming something itch.io would encourage?

**Comment by fasterthanlime** [comment]
@wrsg currently my focus is on getting existing games to run well (whether they're 32-bit or 64-bit builds). For linux, this'll probably mean letting the itch app fetch builds from something like [steam-runtime](https://github.com/valvesoftware/steam-runtime/) if games require them.

Getting current builds to run is the low-hanging fruit - as for encouraging better packaging practices, it requires a bit of backend work we haven't done yet, to analyze the contents of all files being uploaded. It's something I really want to do in the future, but it's quite involved :)
